### PR TITLE
apiextensions: [alpha] opt-in CR pruning according to OpenAPI schema

### PR DESCRIPTION
**Pull request by sttts**
Wire in OpenAPI schema based pruning into
- storage decoder
- request deserializer (which is used for incoming request, but also conversion during patch and admission webhooks)

Follow-up of https://github.com/kubernetes/kubernetes/pull/63604, with pruning only.

```release-note
Add CustomResource pruning as opt-in via spec.prune as alpha feature, available via the feature gate `CustomResourcePruning` (defaults to off). If enabled for a CustomResourceDefinition all fields not specified in the OpenAPI schema are dropped when received via a request, or read from etcd. A sub-object can be excluded from pruning by setting the OpenAPI x-kubernetes-no-prune extension to true.
```

### CODE REVIEW

**Comment by erictune** [comment]
Please ensure this supports having a subtree of the object that is not pruned.
E.g. this schema:
```  
spec:
      a: int
      b: string
      c: raw message 
```
Will cause this object:

```
spec:
    a: 1
    b: "2"
    c: 
        thisPropertyDefinitelyNotMentionedInTheSchema: 1
        norThisOne: 
            foo: bar
    d: 4
```
will be pruned to this:
```
spec:
    a: 1
    b: "2"
    c: 
        thisPropertyDefinitelyNotMentionedInTheSchema: 1
        norThisOne: 
            foo: bar
```

**Comment by sttts** [comment]
@erictune thanks for bringing this up. I see two use-cases we have to support:

1. `map[string]SomeTypeWeKnow`, i.e. unknown map keys, but known and prunable types in the values. This should already work with an empty `properties` field and `additionalProperties` for the values. Will add a testcase for this.
2. `map[string]Interface{}` or even generic `interface{}`, i.e. even arrays and base types. We need a way to mark such a value in the spec. The easiest way would be some custom format `JSON`. That should be valid in JSON-Schema / OpenAPI. Or we invent some `x-*` attribute for that. Have to double check whether we have similar extensions already on the kube-openapi side.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 1]
```diff
@@ -54,6 +54,8 @@ type CustomResourceDefinitionSpec struct {
 	Versions []CustomResourceDefinitionVersion `json:"versions,omitempty" protobuf:"bytes,7,rep,name=versions"`
 	// AdditionalPrinterColumns are additional columns shown e.g. in kubectl next to the name. Defaults to a created-at column.
 	AdditionalPrinterColumns []CustomResourceColumnDefinition `json:"additionalPrinterColumns,omitempty" protobuf:"bytes,8,rep,name=additionalPrinterColumns"`
+	// Prune enables pruning of fields not specified in validation.openAPIV3Schema. Note: this will default to true in version v1.
+	Prune bool `json:"prune,omitempty" protobuf:"varint,9,opt,name=prune"`
```
Comment (mvladev) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: Shouldn't this be an optional value (`*bool`)?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: It defaults to false in v1beta1, to true in newly created GA CRDs. So technically we can keep it a non-pointer now.

Comment (mvladev) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: I was more concerned with the api-conventions doc which I linked above

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: If you need to round trip a false value (from a CRD created in v1beta1) in new versions of CRDs, you need `*bool`

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: To agree about the plan, this it is (if we go with `bool`):

1. an object being created via v1beta1 we default to false.
2. an old v1beta1 object read from etcd is defaulted to false.
3. v1 and internal will get a bool pointer
4. an v1 object defaults to `*true` when created (and read from storage, but in storage we know to have it set anyway).
5. when converting from v1beta1 to internal `false` convert to `*false`, `true` to `*true`.
6. when converting from internal to v1beta1, nil never exists (but we can convert to true, `*false` defaults to false, `*true` default to `true`.
7. we keep the omitempty in v1beta1.

Instead, we could use `*bool` in v1 already. It makes the transition easier maybe. @liggitt opinion? you have had fights with similar cases before. Am happy with either way.

Comment (mvladev) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: > Instead, we could use *bool in v1 already.

I think it'll be easier to have `*bool` in both `internal`, `v1` and `v1beta1` making the conversion without a problem and it'll ease developers who develop against `v1beta1` and want to switch to `v1` once its released

in `v1beta1/zz_generated.defaults.go` we might have:

```go
defaultPrune := false
if in.Spec.Prune == nil {
     in.Spec.Prune = &defaultPrune
}
```

and in `v1/zz_generated.defaults.go`:

```go
defaultPrune := true
if in.Spec.Prune == nil {
     in.Spec.Prune = &defaultPrune
}
```

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: at least it is symmetric. I will change it to `*bool` for v1beta1.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: done

**Comment by spiffxp** [comment]
ping, needs rebase

**Comment by spiffxp** [comment]
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/575 which is currently planned for v1.13

**Comment by shubheksha** [comment]
@nikopen, can we remove the milestone from this PR as https://github.com/kubernetes/enhancements/issues/575 isn't tracked anymore.

**Comment by mrshelly** [comment]
??
Thank you , For read this Email !

________________________________
发件人: k8s-ci-robot <notifications@github.com>
发送时间: 2018年7月25日 16:01
收件人: kubernetes/kubernetes
抄送: Subscribed
主题: Re: [kubernetes/kubernetes] apiextensions: [alpha] opt-in CR pruning according to OpenAPI schema (#64558)


@sttts<https://github.com/sttts>: The following tests failed, say /retest to rerun them all:

Test name       Commit  Details Rerun command
pull-kubernetes-bazel-test      287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-bazel-test/53777/>  /test pull-kubernetes-bazel-test
pull-kubernetes-bazel-build     287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-bazel-build/51268/> /test pull-kubernetes-bazel-build
pull-kubernetes-e2e-gce 287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-e2e-gce/45561/>     /test pull-kubernetes-e2e-gce
pull-kubernetes-e2e-kops-aws    287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-e2e-kops-aws/98465/>        /test pull-kubernetes-e2e-kops-aws
pull-kubernetes-typecheck       287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-typecheck/18208/>   /test pull-kubernetes-typecheck
pull-kubernetes-verify  287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-verify/99811/>      /test pull-kubernetes-verify
pull-kubernetes-e2e-gce-100-performance 287dc44<https://github.com/kubernetes/kubernetes/commit/287dc44008aa38b7712e3566ff84fbf63a234fcb>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64558/pull-kubernetes-e2e-gce-100-performance/13910/>     /test pull-kubernetes-e2e-gce-100-performance

Full PR test history<https://k8s-gubernator.appspot.com/pr/64558>. Your PR dashboard<https://k8s-gubernator.appspot.com/pr/sttts>. Please help us cut down on flakes by linking to<https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests> an open issue<https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when you hit one in your PR.

Instructions for interacting with me using PR comments are available here<https://git.k8s.io/community/contributors/guide/pull-requests.md>. If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra<https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:> repository. I understand the commands that are listed here<https://go.k8s.io/bot-commands>.

―

**Comment by liggitt** [comment]
closing in favor of https://github.com/kubernetes/kubernetes/pull/77333

**Pull request closed without merging**